client/order/bookside_test.go

```diff
@@ -0,0 +1,1311 @@
+				ascending,
+			),
+			entry: &Order{
+				OrderID:  [32]byte{'c'},
+				Side:     msgjson.BuyOrderNum,
+				Quantity: 2,
+				Rate:     3,
+				Time:     10,
+			},
+			expected: makeBookSide(
+				map[uint64][]*Order{
+					1: {
+						&Order{
+							OrderID:  [32]byte{'a'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 10,
+							Rate:     1,
+							Time:     10,
+						},
+						&Order{
+							OrderID:  [32]byte{'b'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 5,
+							Rate:     1,
+							Time:     10,
+						},
+					},
+				},
+				makeRateIndex([]uint64{1}),
+				ascending,
+			),
+			wantErr: true,
+		},
+		{
+			side: makeBookSide(
+				map[uint64][]*Order{
+					1: {
+						&Order{
+							OrderID:  [32]byte{'a'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 10,
+							Rate:     1,
+							Time:     10,
+						},
+						&Order{
+							OrderID:  [32]byte{'b'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 5,
+							Rate:     1,
+							Time:     10,
+						},
+						&Order{
+							OrderID:  [32]byte{'c'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 2,
+							Rate:     1,
+							Time:     10,
+						},
+					},
+				},
+				makeRateIndex([]uint64{1}),
+				ascending,
+			),
+			entry: &Order{
+				OrderID:  [32]byte{'b'},
+				Side:     msgjson.BuyOrderNum,
+				Quantity: 5,
+				Rate:     1,
+				Time:     10,
+			},
+			expected: makeBookSide(
+				map[uint64][]*Order{
+					1: {
+						&Order{
+							OrderID:  [32]byte{'a'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 10,
+							Rate:     1,
+							Time:     10,
+						},
+						&Order{
+							OrderID:  [32]byte{'c'},
+							Side:     msgjson.BuyOrderNum,
+							Quantity: 2,
+							Rate:     1,
+							Time:     10,
+						},
+					},
+				},
+				makeRateIndex([]uint64{1}),
+				ascending,
+			),
+			wantErr: false,
+		},
+	}
+
+	for idx, tc := range tests {
+		err := tc.side.Remove(tc.entry)
+		if (err != nil) != tc.wantErr {
+			t.Fatalf("[BookSide.Remove] #%d: error: %v, wantErr: %v",
+				idx+1, err, tc.wantErr)
+		}
+
+		if len(tc.side.bins) != len(tc.expected.bins) {
+			t.Fatalf("[BookSide.Remove] #%d: expected bin size of %d,"+
+				" got %d", idx+1, len(tc.expected.bins),
+				len(tc.side.bins))
+		}
+
+		for price, bin := range tc.side.bins {
+			expected := tc.expected.bins[price]
+			for i := 0; idx < len(bin); idx++ {
+				if !bytes.Equal(bin[i].OrderID[:], expected[i].OrderID[:]) {
+					t.Fatalf("[BookSide.Remove] #%d: expected price bin %d "+
+						"entry at index %d to be with id %x, got %x", idx+1,
+						price, idx, expected[idx].OrderID[:],
+						bin[idx].OrderID[:])
+				}
+			}
+		}
+
+		entryBin := tc.side.bins[tc.entry.Rate]
+		expectedBin := tc.expected.bins[tc.entry.Rate]
+
+		if len(entryBin) != len(expectedBin) {
+			t.Fatalf("[BookSide.Remove] #%d: expected bin size of %d,"+
+				" got %d", idx+1, len(expectedBin), len(entryBin))
+		}
```

Comment:
You'll need to check the order IDs within the rate bin to make sure the orders are sorting properly by timestamp and then order ID. If I'm reading correctly, you aren't yet varying timestamp in your tests to ensure the intended effects on the sort order. That'll need to be checked too.